### ISSUE TITLE
Model created with Icefall recipe for Common Voice is not working with Sherpa-ONNX

### ISSUE THREAD

**Issue by ChrystianKacki** [k2-fsa/icefall#1539]
I've successfully trained a zipformer streaming model with Common Voice dataset using the receipt located in [https://github.com/k2-fsa/icefall/tree/master/egs/commonvoice](https://github.com/k2-fsa/icefall/tree/master/egs/commonvoice).
I used scripts from `commonvoice/ASR/pruned_transducer_stateless7_streaming`.

Unfortunately the model exported to ONNX is not working with the newest, build from source version of Sherpa-ONNX.
It loads properly, but when real time recognizing starts, it outputs some words on silence and rarely some wrong words when talking.
I tried and everything works fine with model `csukuangfj/sherpa-onnx-streaming-zipformer-en-2023-06-21` from [here](https://k2-fsa.github.io/sherpa/onnx/pretrained_models/online-transducer/zipformer-transducer-models.html#csukuangfj-sherpa-onnx-streaming-zipformer-en-2023-06-21-english).
Maybe it is a matter of number of epochs, because I'm training only 3 for now, using single GPU.
Should I post this issue on Icefall or maybe better on Sherpa-ONNX GitHub repository ?

I execute `train.py` in the following way:
```
./pruned_transducer_stateless7_streaming/train.py \
    --world-size 1 \
    --num-epochs 3 \
    --start-epoch 1 \
    --use-fp16 True \
    --max-duration 550 \
    --enable-musan False \
    --bpe-model data/pl/lang_bpe_500/bpe.model \
    --language pl \
    --cv-manifest-dir data/pl/fbank \
    --exp-dir ./exp
```

I execute `export-onnx.py` in the following way:
```
./pruned_transducer_stateless7_streaming/export-onnx.py \
    --epoch 3 \
    --avg 1 \
    --use-averaged-model False \
    --tokens data/pl/lang_bpe_500/tokens.txt \
    --exp-dir ./exp
```

My model config file looks like this:
```
encoder=C:/sherpa-onnx-model/exp/encoder-epoch-3-avg-1.onnx
decoder=C:/sherpa-onnx-model/exp/decoder-epoch-3-avg-1.onnx
joiner=C:/sherpa-onnx-model/exp/joiner-epoch-3-avg-1.onnx
tokens=C:/sherpa-onnx-model/data/pl/lang_bpe_500/tokens.txt
```

**Comment by JinZr** [comment]
hi, it's ok posting it here, fangjun will look into it at his convenience.

@csukuangfj 

best
jin

**Comment by ChrystianKacki** [comment]
hi, thanks!

**Comment by csukuangfj** [comment]
Has your model converged? Have you run decode.py in icefall to test the WER of your model?

**Comment by ChrystianKacki** [comment]
> Has your model converged?

I don't know what **_converging_** means, could you tell me please?

> Have you run decode.py in icefall to test the WER of your model?

I run `decode.py` with command:
```
./pruned_transducer_stateless7_streaming/decode.py \
    --epoch 3 \
    --avg 1 \
    --max-duration 600 \
    --decode-chunk-len 32 \
    --decoding-method greedy_search \
    --use-averaged-model False \
    --enable-musan False \
    --bpe-model data/pl/lang_bpe_500/bpe.model \
    --lang-dir data/pl/lang_bpe_500 \
    --language pl \
    --cv-manifest-dir data/pl/fbank \
    --exp-dir ./exp
```
It gives the WER with value of 99.75. I guess it is very bad and means that there are almost all errors.

I attach logs from decoding:
[log-decode-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1-2024-03-09-13-11-59.txt](https://github.com/k2-fsa/icefall/files/14546262/log-decode-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1-2024-03-09-13-11-59.txt)
[wer-summary-test-cv-greedy_search.txt](https://github.com/k2-fsa/icefall/files/14546263/wer-summary-test-cv-greedy_search.txt)
[recogs-test-cv-greedy_search-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1.txt](https://github.com/k2-fsa/icefall/files/14546264/recogs-test-cv-greedy_search-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1.txt)
[errs-test-cv-greedy_search-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1.txt](https://github.com/k2-fsa/icefall/files/14546265/errs-test-cv-greedy_search-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1.txt)

**Comment by csukuangfj** [comment]
> I don't know what converging means, could you tell me please?

If `decode.py` gives a reasonable WER, then we say the model is converging/converged.

In your case, the WER is 99.75, which means the model has not been converged.

**Comment by csukuangfj** [comment]
By the way, you can check whether the training loss and validation loss are going down.

**Comment by ChrystianKacki** [comment]
> By the way, you can check whether the training loss and validation loss are going down.

The loss is going down in `train.py` log: [log-train-2024-03-09-12-27-06.txt](https://github.com/k2-fsa/icefall/files/14546355/log-train-2024-03-09-12-27-06.txt)

The `decode.py` log shows that recognition outputs only few words: 'DLATEGO', 'NA', 'PO', 'PRZE', 'W', 'Z'. You can see it attached below:
[errs-test-cv-greedy_search-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1.txt](https://github.com/k2-fsa/icefall/files/14546360/errs-test-cv-greedy_search-epoch-3-avg-1-streaming-chunk-size-32-context-2-max-sym-per-frame-1.txt).

In my application, which uses Sherpa-ONNX DLL libraries built from newest sources, the same words are printed out instead of correct sentences when talking to microphone. Also the model `csukuangfj/sherpa-onnx-streaming-zipformer-en-2023-06-21 (English)` works perfectly in my application and microphone.

**Comment by csukuangfj** [comment]
Your loss is still very very high. If your loss goes below 0.08, you can try to use sherpa-onnx.

Your issue is that your model has not been converged and it is not able to recongize correctly.

Not every model is able to recognize speech correctly. You have to understand that training is very important.
You cannot use a model and expect it to recognize speech accurately.

**Comment by ChrystianKacki** [comment]
Now I understand.
I will try to train my model below loss 0.08.
To do this I should increase the number of epochs when training, am I right?

**Comment by csukuangfj** [comment]
How much data do you have?

Your loss is still 0.8 at epoch 3 and it is very likely that it won't go below 0.1 if you try to train it for more epochs.

**Comment by ChrystianKacki** [comment]
It is Common Voice PL (Polish) model, release 16.1-2023-12-06. Unpacked it occupies 4,42 GB and consists 138 927 audio clips.

**Comment by pzelasko** [comment]
Polish CV is fairly small with <200h, you might want to try adding Polish part of MLS to the mix, and other Polish speech data you can get your hands on. It also helps a lot if you initialize the training for an existing trained model (e.g. English or multilingual checkpoint that was trained on lots of data).

**Comment by ChrystianKacki** [comment]
@pzelasko Thanks for the information! I have a question: what 'MLS' shortcut stands for ? I couldn't find it on the web.

**Comment by ChrystianKacki** [comment]
@JinZr Great, thanks!

**Comment by ChrystianKacki** [comment]
> If your loss goes below 0.08, you can try to use sherpa-onnx.

I guess that 0.08 is an absolute minimum to approach.
What value the training loss and validation loss should have to stop training ?

**Comment by JinZr** [comment]
Nope, empirically speaking 0.08 is where your model gets an acceptable
performance. Also the performance of an asr model doesn’t necessarily
reflected by the value of trining loss or validation loss, that’s what my
experience is. I would say trying at least 30 epochs to start with, and see
if further performance improvements can be obtained by training more
epochs.

Best Regards
Jin

**Comment by ChrystianKacki** [comment]
@JinZr Thanks, Best Regards.
